translated/tech/Write a Kubernetes-ready service from zero step-by-step.md

```diff
@@ -0,0 +1,780 @@
+	r := mux.NewRouter()
+	r.HandleFunc("/home", home).Methods("GET")
+	return r
+}
+```
+
+`handlers/home.go`
+
+```go
+package handlers
+
+import (
+	"fmt"
+	"net/http"
+)
+
+// home is a simple HTTP handler function which writes a response.
+func home(w http.ResponseWriter, _ *http.Request) {
+	fmt.Fprint(w, "Hello! Your request was processed.")
+}
+```
+
+然后`main.go`中做点小改动：
+
+```go
+package main
+
+import (
+	"log"
+	"net/http"
+
+	"github.com/rumyantseva/advent-2017/handlers"
+)
+
+// How to try it: go run main.go
+func main() {
+	log.Print("Starting the service...")
+	router := handlers.Router()
+	log.Print("The service is ready to listen and serve.")
+	log.Fatal(http.ListenAndServe(":8000", router))
+}
+```
+
+## 第四步 添加测试
+
+这一步要开始加点测试了。我们用到了`httptest`包。`Router`方法的测试代码如下：
+
+`handlers/handlers_test.go`：
+
+```go
+package handlers
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+func TestRouter(t *testing.T) {
+	r := Router()
+	ts := httptest.NewServer(r)
+	defer ts.Close()
+
+	res, err := http.Get(ts.URL + "/home")
+	if err != nil {
+		t.Fatal(err)
+	}
+	if res.StatusCode != http.StatusOK {
+		t.Errorf("Status code for /home is wrong. Have: %d, want: %d.", res.StatusCode, http.StatusOK)
+	}
+
+	res, err = http.Post(ts.URL+"/home", "text/plain", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if res.StatusCode != http.StatusMethodNotAllowed {
+		t.Errorf("Status code for /home is wrong. Have: %d, want: %d.", res.StatusCode, http.StatusMethodNotAllowed)
+	}
+
+	res, err = http.Get(ts.URL + "/not-exists")
+	if err != nil {
+		t.Fatal(err)
+	}
+	if res.StatusCode != http.StatusNotFound {
+		t.Errorf("Status code for /home is wrong. Have: %d, want: %d.", res.StatusCode, http.StatusNotFound)
+	}
+}
+```
+
+检查了`GET`请求`/home`路径是否返回`200`，而`POST`请求该路径应该要返回`405`。请求不存在的路由期望返回`404`。实际上，这样子测有点太冗余了，`gorilla/mux`中已经包含类似的测试，所以测试代码可以简化下。
+
+对于`home`来说，检查其返回得code和body值即可。
+
+`handlers/home_test.go`
+
+```go
+package handlers
+
+import (
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+func TestHome(t *testing.T) {
+	w := httptest.NewRecorder()
+	home(w, nil)
+
+	resp := w.Result()
+	if have, want := resp.StatusCode, http.StatusOK; have != want {
+		t.Errorf("Status code is wrong. Have: %d, want: %d.", have, want)
+	}
+
+	greeting, err := ioutil.ReadAll(resp.Body)
+	resp.Body.Close()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if have, want := string(greeting), "Hello! Your request was processed."; have != want {
+		t.Errorf("The greeting is wrong. Have: %s, want: %s.", have, want)
+	}
+}
+```
+
+运行`go test`开始测试。
+
+```
+$ go test -v ./...
+?       github.com/rumyantseva/advent-2017      [no test files]
+=== RUN   TestRouter
+--- PASS: TestRouter (0.00s)
+=== RUN   TestHome
+--- PASS: TestHome (0.00s)
+PASS
+ok      github.com/rumyantseva/advent-2017/handlers     0.018s
+```
+
+## 第五步 添加配置
+
+服务需要能够可配置。前面代码，写死监听`8000`端口，后面需要改成可配置。[The Twelve-Factor App manifesto](https://12factor.net/)，这篇文章详尽阐述了如何写好服务，文中提倡在环境变量中存放配置。后面代码展示如何利用环境变量：
```

Comment:
原文 “and probably it might be useful to be able to configure this value” 翻译不准确。
个人认为全文的类似问题都是因为你的语言习惯。我们的读者很多可能是刚入学的学生，或者是之前还没什么技术背景的人，当然也有可能是技术大牛。为了照顾各方面的需要，建议还是更加按照原文的语气语意翻译比较好。比如上文中有出现过的直接将疑问句翻译成陈述句，这里的probably直接翻译成“需要改成可配置”等等。相信译者肯定是个有足够技术背景，然后对原文理解又非常到位的人，所以才在理解的基础上多了自己的表达，但为了最大限度地照顾读者，希望还是翻译得更加贴切一点为好 ;)

Reply:
OK，辛苦点评了~👍。“按照原文的语气语意翻译”建议后面的翻译中我会注意。